Port FakeProcessGroup to cpp

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118426
* #113057

### Summary
Native functional collective ops requires the backend to be implemented in C++. Porting `FakeProcessGroup` to cpp so that it can also work for native functional collective ops.

The existing tests involving `FakeProcessGroup` all pass. In addition, `DeviceMeshTest::test_fake_pg_device_mesh` now pass with `_USE_NATIVE_C10D_FUNCTIONAL=1`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
Port FakeProcessGroup to cpp

[ghstack-poisoned]

Commit:
Update on "Port FakeProcessGroup to cpp"


### Summary
Native functional collective ops requires the backend to be implemented in C++. Porting `FakeProcessGroup` to cpp so that it can also work for native functional collective ops.

The existing tests involving `FakeProcessGroup` all pass. In addition, `DeviceMeshTest::test_fake_pg_device_mesh` now pass with `_USE_NATIVE_C10D_FUNCTIONAL=1`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented:
awesome! cc @zdevito

wconstab commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2670,6 +2671,13 @@ such as `dist.all_reduce(tensor, async_op=True)`.
         return ivalue.toCustomClass<::c10d::Work>();
       });
 
+  auto fakeProcessGroup =
```
wow is this all the binding work needed?

do you need to add some .pyi stuff and docs?

yifuwang replied:
> wow is this all the binding work needed?

Yup. `torch::CustomClassHolder` is doing the heavy lifting here :)

> do you need to add some .pyi stuff and docs?

Nope. It is seen as `torch.distributed.distributed_c10d.ProcessGroup` in Python which is already covered.

```python
>>> import torch.distributed as dist
>>> from torch.testing._internal.distributed.fake_pg import FakeStore
>>> store = FakeStore()
>>> dist.init_process_group("fake", rank=0, world_size=2, store=store)
>>> print(dist.group.WORLD)
<torch.distributed.distributed_c10d.ProcessGroup object at 0x7f6a0df9ff70>
```

wconstab reviewed (commented):
This is exciting. we should probably use this for more stuff.  I wonder if we have a swath of tests we could rewrite to use this and save CI time.

Also thinking whether we should start having derived classes of this where we can teach it to produce FakeWork's that have some particular behavior

Commit:
Update on "Port FakeProcessGroup to cpp"


### Summary
Native functional collective ops requires the backend to be implemented in C++. Porting `FakeProcessGroup` to cpp so that it can also work for native functional collective ops.

The existing tests involving `FakeProcessGroup` all pass. In addition, `DeviceMeshTest::test_fake_pg_device_mesh` now pass with `_USE_NATIVE_C10D_FUNCTIONAL=1`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (commented):
Porting FakePG to C++ sounds good to me! However I actually don't understand fully why we need FakeProcessGroup be in C++ for our new native functional collective to work, could you elaborate more in the PR summary?

I wondered is it mainly because our new functional c10d collective would call some `resolve_process_group` API that would only work if the process group is registered in the C++ land? If so do you think if we make the python PG creation register to the C++ pg table it would resolve the problem?

yifuwang commented:
> However I actually don't understand fully why we need FakeProcessGroup be in C++ for our new native functional collective to work, could you elaborate more in the PR summary?

> I wondered is it mainly because our new functional c10d collective would call some resolve_process_group API that would only work if the process group is registered in the C++ land?

@wanchaol This is correct. Native functional collective ops require the backend to be implemented in C++.

> If so do you think if we make the python PG creation register to the C++ pg table it would resolve the problem?

It's not just about the registration but also the implementations. In fact, we can register the python `FakeProcessGroup` via the `register_process_group`. However on the C++ side, only the base class impls are visible which are "not implemented" stubs.

I think it should be possible to dispatch back to Python impls if the need arises. For now I think it makes more sense to just port `FakeProcessGroup` to c++.

wanchaol reviewed (approved):
looks great!

wconstab commented:
> @wconstab For most cases though, FakePG won't work with real data right? IIUC doesn't that make it less useful in testing?

right, we can use MultiThreadedProcessGroup for cases where we need real data.

Commit:
Update on "Port FakeProcessGroup to cpp"


### Summary
Native functional collective ops requires the backend to be implemented in C++. Porting `FakeProcessGroup` to cpp so that it can also work for native functional collective ops.

The existing tests involving `FakeProcessGroup` all pass. In addition, `DeviceMeshTest::test_fake_pg_device_mesh` now pass with `_USE_NATIVE_C10D_FUNCTIONAL=1`.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
